README.md
CPUMiner-Multi
==============

[![Build Status](https://travis-ci.org/tpruvot/cpuminer-multi.svg)](https://travis-ci.org/tpruvot/cpuminer-multi)

This is a multi-threaded CPU miner,
fork of [pooler](//github.com/pooler)'s cpuminer (see AUTHORS for list of contributors).

\* __Note__

I (tbauer516) forked and merged the previous versions of the file.
Forked from: lucasjones -> tpruvot -> dan-and -> tbauer516
Merged: m7m into linux, linux-aarch64 into linux

Now includes an m7m hash for Magicoin and the build.sh file now works for most sbc's

#### Table of contents

* [Algorithms](#algorithms)
* [Dependencies](#dependencies)
* [Download](#download)
* [Build](#build)
* [Usage instructions](#usage-instructions)
* [Donations](#donations)
* [Credits](#credits)
* [License](#license)

Algorithms
==========
#### Currently supported
 * ✓ __scrypt__ (Litecoin, Dogecoin, Feathercoin, ...)
 * ✓ __scrypt:N__
 * ✓ __scrypt-jane:N__
 * ✓ __sha256d__ (Bitcoin, Freicoin, Peercoin/PPCoin, Terracoin, ...)
 * ✓ __axiom__ (Axiom Shabal-256 based MemoHash)
 * ✓ __bastion__ (Joincoin [J])
 * ✓ __bitcore__ Permuted serie of 10 algos (BitCore)
 * ✓ __blake__ (Saffron [SFR] Blake-256)
 * ✓ __blake2s__ (NevaCoin Blake2-S 256)
 * ✓ __bmw__ (Midnight [MDT] BMW-256)
 * ✓ __cryptonight__ (Bytecoin [BCN], Monero [XMR])
 * ✓ __cryptonight-light__ (Aeon)
 * ✓ __decred__ (Blake256-14 [DCR])
 * ✓ __dmd-gr__ (Diamond-Groestl)
 * ✓ __fresh__ (FreshCoin)
 * ✓ __groestl__ (Groestlcoin)
 * ✓ __jha__ (JackpotCoin, SweepStake)
 * ✓ __lbry__ (LBRY Credits [LBC])
 * ✓ __lyra2RE__ (Cryptocoin)
 * ✓ __lyra2REv2__ (VertCoin [VTC])
 * ✓ __myr-gr__ Myriad-Groestl (MyriadCoin [MYR])
 * ✓ __m7m__ (Magicoin [XMG])
 * ✓ __neoscrypt__ (Feathercoin)
 * ✓ __nist5__ (MistCoin [MIC], TalkCoin [TAC], ...)
 * ✓ __pentablake__ (Joincoin)
 * ✓ __pluck__ (Supcoin [SUP])
 * ✓ __quark__ (Quarkcoin)
 * ✓ __qubit__ (GeoCoin)
 * ✓ __skein__ (Skeincoin, Myriadcoin, Xedoscoin, ...)
 * ✓ __skein2__ (Woodcoin)
 * ✓ __s3__ (OneCoin)
 * ✓ __sia__ (Reversed Blake2B for SIA [SC])
 * ✓ __sib__ X11 + gost streebog (SibCoin)
 * ✓ __timetravel__ Permuted serie of 8 algos (MachineCoin [MAC])
 * ✓ __tribus__ 3 of the top NIST5 algos (Denarius [DNR])
 * ✓ __vanilla__ (Blake-256 8-rounds - double sha256 [VNL])
 * ✓ __veltor__ (Veltor [VLT])
 * ✓ __xevan__ x17 x 2 on bigger header (BitSend [BSD])
 * ✓ __x11evo__ (Revolver [XRE])
 * ✓ __x11__ (Darkcoin [DRK], Hirocoin, Limecoin, ...)
 * ✓ __x13__ (Sherlockcoin, [ACE], [B2B], [GRC], [XHC], ...)
 * ✓ __x14__ (X14, Webcoin [WEB])
 * ✓ __x15__ (RadianceCoin [RCE])
 * ✓ __x17__ (Verge [XVG])
 * ✓ __yescrypt__ (GlobalBoostY [BSTY], Unitus [UIS], MyriadCoin [MYR])
 * ✓ __zr5__ (Ziftrcoin [ZRC])

#### Implemented, but untested
 * ? hefty1 (Heavycoin)
 * ? keccak (Maxcoin  HelixCoin, CryptoMeth, Galleon, 365coin, Slothcoin, BitcointalkCoin)
 * ? luffa (Joincoin, Doomcoin)
 * ? shavite3 (INKcoin)

#### Planned support for
 * *scrypt-jane* (YaCoin, CopperBars, Pennies, Tickets, etc..)
 
Dependencies
============
 * libcurl http://curl.haxx.se/libcurl/
 * jansson http://www.digip.org/jansson/ (jansson source is included in-tree)
 * openssl libcrypto https://www.openssl.org/
 * libgmp (GNU Multiple Precision Library) https://gmplib.org/
 * pthreads
 * zlib (for curl/ssl)

Download
========
 * Windows releases: https://github.com/tpruvot/cpuminer-multi/releases
 * Git tree:   https://github.com/tpruvot/cpuminer-multi
   * Clone with `git clone https://github.com/tpruvot/cpuminer-multi`

Build
=====

#### Basic *nix build instructions:
 * just use `./build.sh`
_OR_

```
 ./autogen.sh	# only needed if building from git repo
 ./nomacro.pl	# only needed if building on Mac OS X or with Clang
 ./configure CFLAGS="*-march=native*" --with-crypto --with-curl
 # Use -march=native if building for a single machine
 make
```

#### Note for Debian/Ubuntu users:

```
 apt-get install automake autoconf pkg-config libcurl4-openssl-dev libjansson-dev libssl-dev libgmp-dev libz-dev make g++
```

#### Note for pi64 users:

```
 ./configure --disable-assembly CFLAGS="-Ofast -march=native" --with-crypto --with-curl
```

#### Notes for AIX users:
 * To build a 64-bit binary, export OBJECT_MODE=64
 * GNU-style long options are not supported, but are accessible via configuration file

#### Basic Windows build with Visual Studio 2013
 * All the required .lib files are now included in tree (windows only)
 * For visual studio, the GNU MP library is available in the project MPIR
 * AVX enabled by default for x64 platform (AVX2 and XOP could also be used)

#### Basic Windows build instructions, using MinGW64:
 * Install MinGW64 and the MSYS Developer Tool Kit (http://www.mingw.org/)
   * Make sure you have mstcpip.h in MinGW\include
 * install pthreads-w64
 * Install libcurl devel (http://curl.haxx.se/download.html)
   * Make sure you have libcurl.m4 in MinGW\share\aclocal
   * Make sure you have curl-config in MinGW\bin
 * Install openssl devel (https://www.openssl.org/related/binaries.html)
 * Download and "make install" sources of libgmp (v5.1 and v6 are ok)
 * In the MSYS shell, run:
   * for 64bit, you can use `./mingw64.sh` else :
     `./autogen.sh	# only needed if building from git repo`
   ```
   LIBCURL="-lcurldll" ./configure CFLAGS="*-march=native*"
   # Use -march=native if building for a single machine
   make
    ```

#### Architecture-specific notes:
 * ARM:
   * No runtime CPU detection. The miner can take advantage of some instructions specific to ARMv5E and later processors, but the decision whether to use them is made at compile time, based on compiler-defined macros.
   * To use NEON instructions, add `"-mfpu=neon"` to CFLAGS.
 * x86:
   * The miner checks for SSE2 instructions support at runtime, and uses them if they are available.
 * x86-64:	
   * The miner can take advantage of AVX, AVX2 and XOP instructions, but only if both the CPU and the operating system support them.
     * Linux supports AVX starting from kernel version 2.6.30.
     * FreeBSD supports AVX starting with 9.1-RELEASE.
     * Mac OS X added AVX support in the 10.6.8 update.
     * Windows supports AVX starting from Windows 7 SP1 and Windows Server 2008 R2 SP1.
   * The configure script outputs a warning if the assembler doesn't support some instruction sets. In that case, the miner can still be built, but unavailable optimizations are left off.

Usage instructions
==================
Run "cpuminer --help" to see options.

### Connecting through a proxy

Use the --proxy option.

To use a SOCKS proxy, add a socks4:// or socks5:// prefix to the proxy host  
Protocols socks4a and socks5h, allowing remote name resolving, are also available since libcurl 7.18.0.

If no protocol is specified, the proxy is assumed to be a HTTP proxy.  
When the --proxy option is not used, the program honors the http_proxy and all_proxy environment variables.

Donations
=========
Donations for the work done in this fork are accepted :

Tanguy Pruvot :
* BTC: `1FhDPLPpw18X4srecguG3MxJYe4a1JsZnd`
* ZRC: `ZX6LmrCwphNgitxvDnf8TX6Tsegfxpeozx`

Lucas Jones :
* MRO: `472haywQKoxFzf7asaQ4XKBc2foAY4ezk8HiN63ifW4iAbJiLnfmJfhHSR9XmVKw2WYPnszJV9MEHj9Z5WMK9VCNHaGLDmJ`
* BTC: `139QWoktddChHsZMWZFxmBva4FM96X2dhE`

Credits
=======
CPUMiner-multi was forked from pooler's CPUMiner, and has been started by Lucas Jones.
* [tpruvot](https://github.com/tpruvot) added all the recent features and newer algorythmns
* [Wolf9466](https://github.com/wolf9466) helped with Intel AES-NI support for CryptoNight

License
=======
GPLv2.  See COPYING for details.


NEWS
Version 1.3.3
- Add tribus algo

Version 1.3.2
- Add bitcore algo
- Add jha algo

Version 1.3.1 (Tanguy Pruvot)  26 Jan 2017
- Add timetravel algo
- Add --max-log-rate to limit per-core logs

Version 1.3 (Tanguy Pruvot)
- Add decred algo
- Add lbry algo
- Add sia algo
- Add x11evo algo
- Add x17 and xevan algos
- Add veltor algo
- Enhance Blake2-S
- Stratum benchmarks support
- Show diff and solved blocks by default

Version 1.2 (Tanguy Pruvot)
- Add cryptonight-light (Aeon)
- Add Lyra2REv2 algo (Vertcoin)
- Allow to load a remote config with curl
- Algorithm parameter is now case insensitive
- Drop anime algo (dead coin)
- Add Sib(coin) algo
- Add Bastion algo
- Add Yescrypt and scrypt-jane algos
- Add m7m algo
- Compute and show network diff in pools too
- Compute Shares diff and count solved blocks
- Checkup on arm, tested ok on Tegra K1 (CyanogenMod 12.1)

version 1.1 (Tanguy Pruvot)
- Add basic API remote control (quit/seturl)
- Add GroestlCoin, Diamond and Myriad variants
- Add Pluck algo and fix gbt query crash
- Add ZR5 algo (ZRC) and fix longpoll bug on linux
- Add Luffa algo
- Add Skein2 algo (Double Skein for Woodcoin)
- Add Animecoin algo (Quark variant)
- Add Dropcoin pok algo
- Add BMW-256 (MDT) algo
- Add Axiom algo
- Change some logged strings
- Use all cores by default, not N-1
- Handle a default config to run without params
- add cpu-priority and cpu-affinity options
- add NSIS installer script for windows setup
- Implement background option on windows
- add -m stratum option (diff-multiplier)
- Time limit to allow benchmarks or cron jobs
- Fix Cryptonight stratum support
- Allow to disable extranonce support

version 1.0.9 (Tanguy Pruvot)
- pool extranonce subscribe
- upgrade jansson
- lyra2 algo
- fix for solo mining
- API websocket support

Version 1.0.8 (Tanguy Pruvot)
- API Monitoring Support
- Enhance config values support (int/real/bool)
- Rewrite blake algo (speed x2)

Version 1.0.7 (Tanguy Pruvot)
- Add NIST5 and QUBIT algos
- Show current stratum block height
- Fix wallet solo mining

Version 1.0.6 (Tanguy Pruvot)
- Fix scrypt algo
- More work on VC2013
- Add -f tuning option to test with reduced difficulty
- Add S3 algo

Version 1.0.5 (Tanguy Pruvot)

- Merge remaining v2.4 cpu-miner changes
- Add colored output (disable with --no-color)
- Test and fix blake on NEOS, needs 14 rounds (was 8)
- Add pentablake (5x blake256) (from bitbandi)
- Add neoscrypt
- Windows (VC++ 2013 and MinGW64 build support)
- Enhance --version informations (compiler + lib versions)

Version 1.0.4 (Tanguy Pruvot)

- Add x13 x14 and x15 algos (Sherlockcoin, X14Coin, Webcoin..)
- Add scrypt:N variants (Vertcoin)
- Add fresh algo
- Fix thread khashes/s value output
- Add a configure option --disable-assembly

Version multi 1.0.3 (Lucas Jones)

- Add new algos :
  x11 (Darkcoin [DRK], Hirocoin, Limecoin)
  cryptonight (Bytecoin [BCN], Monero)
  keccak (Maxcoin  HelixCoin, CryptoMeth, Galleon, 365coin, Slothcoin, BitcointalkCoin)
  hefty1 (Heavycoin)
  quark (Quarkcoin)
  skein (Skeincoin, Myriadcoin)
  shavite3 (INKcoin)
  blake (Blakecoin)

- See README.md

Version 2.4 - May 20, 2014

- Add support for the getblocktemplate RPC method (BIP 22)
- Allow tunnelling Stratum through HTTP proxies
- Add a --no-redirect option to ignore redirection requests
- Timeout for long polling is now disabled by default
- Fix CPU affinity on Linux (kiyominer)
- Add support for building under 64-bit Cygwin
- Expand version information with build details

Version 2.3.3 - Feb 27, 2014

- The --url option is now mandatory
- Do not switch to Stratum when using an HTTP proxy
- Fix scheduling policy change on Linux (clbr)
- Fix CPU affinity on FreeBSD (ache)
- Compatibility fixes for various platforms, including Solaris 8
  and old versions of OS X
- A man page for minerd is now available

Version 2.3.2 - Jul 10, 2013

- Add optimizations for AVX2-capable x86-64 processors
- Ensure that the output stream is flushed after every log message
- Fix an undefined-behavior bug in the Stratum code

Version 2.3.1 - Jun 18, 2013

- Add a --cert option for specifying an SSL certificate (martinwguy)
- Fix a bug that only made SHA-256d mining work at difficulty 1
- Fix a couple of compatibility issues with some Stratum servers

Version 2.3 - Jun 12, 2013

- Add support for the Stratum mining protocol
- Automatically switch to Stratum if the mining server supports
  the X-Stratum extension, unless --no-stratum is used
- Set CPU affinity on FreeBSD (lye)
- Fix a bug in libcurl initialization (martinwguy)

Version 2.2.3 - Aug 5, 2012

- Add optimized ARM NEON code for scrypt and SHA-256d
- Add a --benchmark option that allows offline testing
- Support for the X-Reject-Reason extension

Version 2.2.2 - Jun 7, 2012

- Various performance improvements for x86 and x86-64
- Optimize scrypt for ARMv5E and later processors
- Set the priority of miner threads to idle on Windows
- Add an option to start minerd as a daemon on POSIX systems

Version 2.2.1 - May 2, 2012

- Add optimized code for ARM processors
- Support for building on NetBSD and OpenBSD
- Various compatibility fixes for AIX (pontius)

Version 2.2 - Apr 2, 2012

- Add an optimized SHA-256d algorithm, with specialized code
  for x86 and x86-64 and support for AVX and XOP instructions
- Slight performance increase for scrypt on x86 and x86-64

algo/velvet.c
#include "miner.h"

#include <gmp.h> /* user MPIR lib for vstudio */
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>
#include <float.h>
#include <math.h>

#include "crypto/magimath.h"

#include "sha3/sph_bmw.h"
#include "sha3/sph_groestl.h"
#include "sha3/sph_jh.h"
#include "sha3/sph_keccak.h"
#include "sha3/sph_skein.h"
#include "sha3/sph_luffa.h"
#include "sha3/sph_cubehash.h"
#include "sha3/sph_shavite.h"
#include "sha3/sph_simd.h"
#include "sha3/sph_echo.h"
#include "sha3/sph_hamsi.h"
#include "sha3/sph_fugue.h"
#include "sha3/sph_shabal.h"
#include "sha3/sph_whirlpool.h"
#include "sha3/sph_sha2.h"
#include "sha3/sph_haval.h"

#define BITS_PER_DIGIT 3.32192809488736234787
#define EPS (DBL_EPSILON)

#define NM25M 23
#define SW_DIVS 23

static inline void mpz_set_uint256(mpz_t r, uint32_t *u) {
	mpz_import(r, 32 / sizeof(unsigned long), -1, sizeof(unsigned long), -1, 0, u);
}

static inline void mpz_set_uint512(mpz_t r, uint32_t *u) {
	mpz_import(r, 64 / sizeof(unsigned long), -1, sizeof(unsigned long), -1, 0, u);
}

static int is_zero_hash(uint32_t* hash) {
	for (int w=0; w<8; w++)
		if (hash[w]) return false;
	return true;
}

void velvethash(void *output, const void *input)
{
	uint32_t _ALIGN(128) hash[25][16];
	uint32_t _ALIGN(128) echo[16] = { 0 };
	uint32_t _ALIGN(128) finalhash[8] = { 0 };

	const void* ptr = input;
	const uint32_t nnNonce = ((uint32_t*)input)[19];
	const int nnNonce2 = (int) (nnNonce / 2);
	const int sz = 80;

	sph_shabal512_context     ctx_shabal;
	sph_bmw512_context        ctx_bmw512;
	sph_groestl512_context    ctx_groestl;
	sph_jh512_context         ctx_jh;
	sph_keccak512_context     ctx_keccak;
	sph_skein512_context      ctx_skein;
	sph_luffa512_context      ctx_luffa;
	sph_cubehash512_context   ctx_cubehash;
	sph_shavite512_context    ctx_shavite;
	sph_simd512_context       ctx_simd;
	sph_echo512_context       ctx_echo;
	sph_hamsi512_context      ctx_hamsi;
	sph_fugue512_context      ctx_fugue;
	sph_whirlpool_context     ctx_whirlpool;
	sph_sha512_context        ctx_sha2;
	sph_haval256_5_context    ctx_haval;
	sph_shabal256_context	  ctx_shabal256;

	sph_shabal512_init(&ctx_shabal);
	sph_shabal512(&ctx_shabal, ptr, sz);
	sph_shabal512_close(&ctx_shabal, hash[0]);

	sph_groestl512_init(&ctx_groestl);
	sph_groestl512(&ctx_groestl, ptr, sz);
	sph_groestl512_close(&ctx_groestl, hash[3]);

	if (hash[0][0] & 25)
	{
		sph_bmw512_init(&ctx_bmw512);
		sph_bmw512(&ctx_bmw512, ptr, sz);
		sph_bmw512_close(&ctx_bmw512, hash[1]);
	} else {
		memcpy(hash[1], hash[3], 64); // groestl
	}

	memcpy(hash[2], hash[0], 64); // shabal
	memcpy(hash[4], hash[0], 64); // shabal

	bool jhDone;
	if (hash[4][0] & 25) {
		sph_skein512_init(&ctx_skein);
		sph_skein512 (&ctx_skein, ptr, sz);
		sph_skein512_close(&ctx_skein, (void*)hash[5]);
		jhDone = false;
	} else {
		sph_jh512_init(&ctx_jh);
		sph_jh512 (&ctx_jh, ptr, sz);
		sph_jh512_close(&ctx_jh, (void*)hash[5]);
		jhDone = true;
	}

	if (hash[5][0] & 25) {
		if (jhDone)
			memcpy(hash[6], hash[5], 64);
		else {
			sph_jh512_init(&ctx_jh);
			sph_jh512 (&ctx_jh, ptr, sz);
			sph_jh512_close(&ctx_jh, (void*)hash[6]);
		}
	} else {
		if (!jhDone)
			memcpy(hash[6], hash[5], 64);
		else {
			sph_skein512_init(&ctx_skein);
			sph_skein512 (&ctx_skein, ptr, sz);
			sph_skein512_close(&ctx_skein, (void*)hash[6]);
		}
	}

	sph_cubehash512_init(&ctx_cubehash);
	sph_cubehash512 (&ctx_cubehash, ptr, sz);
	sph_cubehash512_close(&ctx_cubehash, (void*)hash[7]);

	sph_shavite512_init(&ctx_shavite);
	sph_shavite512(&ctx_shavite, ptr, sz);
	sph_shavite512_close(&ctx_shavite, (void*)hash[8]);

	if (hash[8][0] & 25) {
		sph_keccak512_init(&ctx_keccak);
		sph_keccak512 (&ctx_keccak, ptr, sz);
		sph_keccak512_close(&ctx_keccak, (void*)hash[9]);
	} else {
		sph_luffa512_init(&ctx_luffa);
		sph_luffa512 (&ctx_luffa, (void*)hash[8], 64);
		sph_luffa512_close(&ctx_luffa, (void*)hash[9]);
	}

	sph_echo512_init(&ctx_echo);
	sph_echo512 (&ctx_echo, ptr, sz);
	sph_echo512_close(&ctx_echo, (void*)echo);

	if (hash[9][0] & 25) {
		sph_simd512_init(&ctx_simd);
		sph_simd512 (&ctx_simd, ptr, sz);
		sph_simd512_close(&ctx_simd, (void*)hash[10]);
	} else {
		memcpy(hash[10], echo, 64);
	}

	// all shabal512
	memcpy(hash[11], hash[0], 64);
	memcpy(hash[12], hash[0], 64);
	memcpy(hash[13], hash[0], 64);
	memcpy(hash[14], hash[0], 64);
	memcpy(hash[15], hash[0], 64);

	if (hash[15][0] & 25) {
		sph_hamsi512_init(&ctx_hamsi);
		sph_hamsi512 (&ctx_hamsi, ptr, sz);
		sph_hamsi512_close(&ctx_hamsi, (void*)hash[16]);
	} else {
		memcpy(hash[16], echo, 64);
	}

	sph_fugue512_init(&ctx_fugue);
	sph_fugue512 (&ctx_fugue, ptr, sz);
	sph_fugue512_close(&ctx_fugue, (void*)hash[17]);

	if (hash[17][0] & 25) {
		memcpy(hash[18], hash[0], 64); // shabal
	} else {
		memcpy(hash[18], echo, 64);
	}

	sph_whirlpool_init(&ctx_whirlpool);
	sph_whirlpool (&ctx_whirlpool, ptr, sz);
	sph_whirlpool_close(&ctx_whirlpool, (void*)hash[19]);

	sph_sha512_init(&ctx_sha2);
	sph_sha512 (&ctx_sha2, ptr, sz);
	sph_sha512_close(&ctx_sha2, (void*)hash[20]);

	sph_haval256_5_init(&ctx_haval);
	sph_haval256_5 (&ctx_haval, ptr, sz);
	memset(&hash[21][8], 0, 32);
	sph_haval256_5_close(&ctx_haval, (void*)hash[21]);

	if (hash[21][0] & 25) {
		memcpy(hash[22], hash[0], 64); // shabal
	} else {
		memcpy(hash[22], echo, 64);
	}

	if (hash[22][0] & 25) {
		memcpy(hash[23], hash[0], 64); // shabal
	} else {
		memcpy(hash[23], echo, 64);
	}

	if (hash[23][0] & 25) {
		memcpy(hash[24], hash[0], 64); // shabal
	} else {
		memcpy(hash[24], echo, 64);
	}

	mpz_t bns[26];

	// Take care of zeros and load gmp
	for(int i=0; i < 25; i++) {
		if (is_zero_hash(hash[i]))
			hash[i][0] = 1;
		mpz_init(bns[i]);
		mpz_set_uint512(bns[i], hash[i]);
	}

	mpz_init(bns[25]);
	mpz_set_ui(bns[25], 0);
	for(int i=0; i < 25; i++)
		mpz_add(bns[25], bns[25], bns[i]);

	mpz_t product;
	mpz_init(product);
	mpz_set_ui(product,1);

	for(int i=0; i < 26; i++)
		mpz_mul(product, product, bns[i]);
	mpz_pow_ui(product, product, 2);

	size_t bytes = mpz_sizeinbase(product, 256);
//	printf("M25M data space: %iB\n", bytes);
	char *data = (char*)malloc(bytes);
	mpz_export(data, NULL, -1, 1, 0, 0, product);

	sph_shabal256_init(&ctx_shabal256);
	sph_shabal256 (&ctx_shabal256, data, bytes);
	sph_shabal256_close(&ctx_shabal256, (void*) finalhash);
	free(data);

	int digits = (int) ((sqrt((double)(nnNonce2))*(1.+EPS))/9000+255);
	int iterations = 20; // <= 500
	mpf_set_default_prec((long int)(digits*BITS_PER_DIGIT+16));

	mpz_t magipi;
	mpz_t magisw;
	mpf_t magifpi;
	mpf_t mpa1, mpb1, mpt1, mpp1;
	mpf_t mpa2, mpb2, mpt2, mpp2;
	mpf_t mpsft;

	mpz_init(magipi);
	mpz_init(magisw);
	mpf_init(magifpi);
	mpf_init(mpsft);
	mpf_init(mpa1);
	mpf_init(mpb1);
	mpf_init(mpt1);
	mpf_init(mpp1);

	mpf_init(mpa2);
	mpf_init(mpb2);
	mpf_init(mpt2);
	mpf_init(mpp2);

	uint32_t usw_;
	usw_ = sw_(nnNonce2, SW_DIVS);
	if (usw_ < 1) usw_ = 1;

	mpz_set_ui(magisw, usw_);
	size_t mpzscale = mpz_size(magisw);
	for(int i=0; i < NM25M; i++)
	{
		if (mpzscale > 1000) mpzscale = 1000;
		else if (mpzscale < 1) mpzscale = 1;

		mpf_set_ui(mpa1, 1);
		mpf_set_ui(mpp1, 1);
		mpf_set_ui(mpb1, 2);
		mpf_sqrt(mpb1, mpb1);
		mpf_ui_div(mpb1, 1, mpb1);
		mpf_set_ui(mpsft, 10);
		mpf_set_d(mpt1, 0.25*mpzscale);

		for(int it=0; it <= iterations; it++)
		{
			mpf_add(mpa2, mpa1, mpb1);
			mpf_div_ui(mpa2, mpa2, 2);
			mpf_mul(mpb2, mpa1, mpb1);
			mpf_abs(mpb2, mpb2);
			mpf_sqrt(mpb2, mpb2);
			mpf_sub(mpt2, mpa1, mpa2);
			mpf_abs(mpt2, mpt2);
			mpf_sqrt(mpt2, mpt2);
			mpf_mul(mpt2, mpt2, mpp1);
			mpf_sub(mpt2, mpt1, mpt2);
			mpf_mul_ui(mpp2, mpp1, 2);
			mpf_swap(mpa1, mpa2);
			mpf_swap(mpb1, mpb2);
			mpf_swap(mpt1, mpt2);
			mpf_swap(mpp1, mpp2);
		}

		mpf_add(magifpi, mpa1, mpb1);
		mpf_pow_ui(magifpi, magifpi, 2);
		mpf_div_ui(magifpi, magifpi, 4);
		mpf_abs(mpt1, mpt1);
		mpf_div(magifpi, magifpi, mpt1);

		mpf_pow_ui(mpsft, mpsft, digits/2);
		mpf_mul(magifpi, magifpi, mpsft);
		mpz_set_f(magipi, magifpi);

		mpz_add(product, product, magipi);
		mpz_add(product, product, magisw);

		if (is_zero_hash(finalhash)) finalhash[0] = 1;
		mpz_set_uint256(bns[0], finalhash);
		mpz_add(bns[25], bns[25], bns[0]);

		mpz_mul(product, product, bns[25]);
		mpz_cdiv_q(product, product, bns[0]);
		if (mpz_sgn(product) <= 0) mpz_set_ui(product,1);

		bytes = mpz_sizeinbase(product, 256);
		mpzscale = bytes;
	//	printf("M25M data space: %iB\n", bytes);

		char *bdata = (char*) malloc(bytes);
		if (!bdata) {
			applog(LOG_ERR, "velvet mem alloc problem!");
			memset(finalhash, 0xff, 32);
			break;
		}
		mpz_export(bdata, NULL, -1, 1, 0, 0, product);

		sph_shabal256_init(&ctx_shabal256);
		sph_shabal256 (&ctx_shabal256, bdata, bytes);
		sph_shabal256_close(&ctx_shabal256, (void*)(&finalhash));

		free(bdata);
	//	applog(LOG_DEBUG, "finalhash:");
	//	applog_hash((char*)finalhash);
	}

	// Free gmp memory
	for(int i=0; i < 26; i++)
		mpz_clear(bns[i]);

	mpz_clear(product);

	mpz_clear(magipi);
	mpz_clear(magisw);
	mpf_clear(magifpi);
	mpf_clear(mpsft);
	mpf_clear(mpa1);
	mpf_clear(mpb1);
	mpf_clear(mpt1);
	mpf_clear(mpp1);

	mpf_clear(mpa2);
	mpf_clear(mpb2);
	mpf_clear(mpt2);
	mpf_clear(mpp2);

	memcpy(output, finalhash, 32);
}

int scanhash_velvet(int thr_id, struct work *work, uint64_t max_nonce, uint64_t *hashes_done)
{
	uint32_t _ALIGN(64) vhash[8];
	uint32_t _ALIGN(64) endiandata[20];
	uint32_t *pdata = work->data;
	uint32_t *ptarget = work->target;

	const uint32_t Htarg = ptarget[7];
	const uint32_t first_nonce = pdata[19];

	uint32_t n = first_nonce;

	for (int i=0; i < 19; i++) {
		be32enc(&endiandata[i], pdata[i]);
	}

	do {
		be32enc(&endiandata[19], n);
		velvethash((char*)vhash, (char*)endiandata);
		if (vhash[7] < Htarg && fulltest(vhash, ptarget)) {
			work_set_target_ratio(work, vhash);
			*hashes_done = n - first_nonce + 1;
			pdata[19] = n;
			return true;
		}
		n++;

	} while (n < max_nonce && !work_restart[thr_id].restart);

	*hashes_done = n - first_nonce + 1;
	pdata[19] = n;

	return 0;
}